<launch>

  <arg name="model"           default="QuadrotorAlpha" />
  <arg name="mass"            default="0.551"/>
  <arg name="speed"           default="000"/>
  <arg name="height"		  default="010"/>
  <!-- speed values: {000, 050, 100, 150, 200, 300, 400, 500, 600, 700, 800} cm/s -->
  <!-- height values: {005, 010, 015, 020, 025, 030, 040, 050} cm -->

  <group ns="$(arg model)">

    <param name="mass" value="$(arg mass)"/>

    <rosparam file="$(find ground_effect)/config/hummingbird_gains.yaml"/>

    <!-- <param name="traj_filename" type="string" value="/home/odroid/perch_data/speed_$(arg speed)cmps/height_$(arg height)cm/traj.csv"/>
    -->

    <node pkg="nodelet"
      type="nodelet"
      name="standalone_nodelet"
      args="manager"
      output="screen"/>

    <!-- Controller -->
    <node pkg="nodelet"
      type="nodelet"
      args="load so3_control/SO3ControlNodelet standalone_nodelet"
      name="so3_control"
      required="true"
      output="screen">
      <rosparam file="$(find ground_effect)/config/hummingbird_gains.yaml"/>
      <param name="mass" value="$(arg mass)"/>
      <remap from="~odom" to="odom"/>
      <remap from="~position_cmd" to="position_cmd"/>
      <remap from="~so3_cmd" to="so3_cmd"/>
      <remap from="~motors" to="motors"/>
    </node>

    <!-- Trackers Manager -->
    <node pkg="nodelet"
      type="nodelet"
      name="trackers_manager"
      args="load trackers_manager/TrackersManager standalone_nodelet"
      output="screen">
      <rosparam file="$(find trackers_manager)/config/trackers.yaml"/>
      <rosparam file="$(find ground_effect)/config/hummingbird_gains.yaml"/>
      <remap from="~odom" to="odom"/>
      <remap from="~cmd" to="position_cmd"/>
    </node>

    <!-- Information transmitted to low level -->
    <node pkg="nodelet"
      type="nodelet"
      name="quad_encode_msg"
      args="load quad_encode_msg/QuadEncodeMsg standalone_nodelet"
      output="screen">
      <remap from="~so3_cmd" to="so3_cmd"/>
      <remap from="~serial_msg" to="to_robot"/>
    </node>

    <!-- Information received from low level -->
    <node pkg="nodelet"
      type="nodelet"
      name="quad_decode_msg"
      args="load quad_decode_msg/QuadDecodeMsg standalone_nodelet"
      output="screen">
      <remap from="~serial" to="from_robot"/>
    </node>

    <!-- For serial communication -->
    <node pkg="nodelet"
      type="nodelet"
      name="quad_serial_comm"
      args="load quad_serial_comm/QuadSerialComm standalone_nodelet"
      output="screen">
      <param name="device" value="/dev/ttyS2"/>
      <param name="baud_rate" value="230400"/>
      <remap from="~to_robot" to="to_robot"/>
      <remap from="~from_robot" to="from_robot"/>
    </node>

    <!-- This is the state control node -->
    <node pkg="ground_effect"
      type="state_control"
      name="state_control"
      output="screen">
      <param name="traj_filename" type="string" value="/home/odroid/perch_data/speed_$(arg speed)cmps/height_$(arg height)cm/traj.csv"/>
      <rosparam file="$(find ground_effect)/config/mav_manager_params.yaml" />
      <remap from="~so3_cmd" to="so3_cmd"/>
    </node>

    <!-- This is the recording node -->
    <node pkg="rosbag" 
      type="record"
      output="screen" 
    	name="rosbag_record"
      args="record -o /home/odroid/perch_data/speed_$(arg speed)cmps/height_$(arg height)cm/ 
      /QuadrotorAlpha/motors /QuadrotorAlpha/odom /QuadrotorAlpha/position_cmd /QuadrotorAlpha/quad_decode_msg/output_data /QuadrotorAlpha/so3_cmd /QuadrotorAlpha/traj_num /QuadrotorAlpha/traj_signal /QuadrotorAlpha/start_time /nanokontrol2"
    /> 
   
  </group>
</launch>
